// react-marquee-line/dist/vertical/ItemBox.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _Item = _interopRequireDefault(require("./Item"));

require("./index.css");

var _helpers = require("./helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var ItemBox = function ItemBox(props) {
  var subList = props.subList,
      onSlideUpEnd = props.onSlideUpEnd;

  var _useState = (0, _react.useState)(0),
      _useState2 = _slicedToArray(_useState, 2),
      y = _useState2[0],
      setY = _useState2[1];

  var _useState3 = (0, _react.useState)(1),
      _useState4 = _slicedToArray(_useState3, 2),
      opacity = _useState4[0],
      setOpacity = _useState4[1];

  var itemBoxRef = (0, _react.useRef)(null);
  var rafRef = (0, _react.useRef)(null);
  var fadeUp = (0, _react.useCallback)(function () {
    rafRef.current && cancelAnimationFrame(rafRef.current);
    var node = itemBoxRef.current;

    function run() {
      if (!node) return false;
      var y = (0, _helpers.getTranslateY)(node.style.transform);

      if (y > -110) {
        setY(function (prev) {
          return prev - 5;
        });
        setOpacity(function (prev) {
          var tmp = prev - 0.05;
          return tmp < 0 ? 0 : tmp;
        });
        rafRef.current = requestAnimationFrame(fadeUp);
      } else {
        onSlideUpEnd();
      }
    }

    run();
  }, [onSlideUpEnd]);
  (0, _react.useEffect)(function () {
    var token = setTimeout(function () {
      fadeUp();
    }, 2000);
    return function () {
      token && clearTimeout(token);
    };
  }, [fadeUp]);
  return _react.default.createElement("div", {
    ref: itemBoxRef,
    className: "react-marquee-line-ver react-marquee-line-ver-itemBox",
    style: {
      transform: "translateY(".concat(y, "%)"),
      opacity: opacity
    }
  }, subList.map(function (item, idx) {
    return _react.default.createElement(_Item.default, {
      key: idx
    }, item);
  }));
};

var _default = _react.default.memo(ItemBox);

exports.default = _default;
